session.go

```diff
@@ -198,6 +198,22 @@ func (s *Session) init() error {
 	}
 	s.ring.endpoints = hosts
 
+	// filter hosts at the onset to avoid using using filtered hosts in the control connection while
+	// discovering the protocol and/or connecting.
+	if s.cfg.HostFilter != nil {
+		newHosts := make([]*HostInfo, 0, len(hosts))
```

Comment:
We can refactor to something like:
```
newHosts := hosts[:0]
```

to avoid allocating a new slice, but in general, it is a good suggestion to fix an issue.